status-group-chats-spec.md

```diff
@@ -28,177 +28,105 @@ This documents describes the group chat protocol used by the status application.
+message MembershipUpdateMessage {
+  // The chat id of the private group chat
+  string chat_id = 1;
+  // A list of events for this group chat, first x bytes are the signature, then is a 
+  // protobuf encoded MembershipUpdateEvent
+  repeated bytes events = 2;
+  // An optional chat message
+  ChatMessage message = 3;
+}
 
-```
-{
-  "events": [struct {"type": string, "member": string, "members": [string], "clock-value": uint, "name": string],
-  "signature": string,
-  "chat-id": string
+message MembershipUpdateEvent {
+  // Lamport timestamp of the event
+  uint64 clock = 1;
+  // List of public keys of objects of the action
+  repeated string members = 2;
+  // Name of the chat for the CHAT_CREATED/NAME_CHANGED event types
+  string name = 3;
+  // The type of the event
+  EventType type = 4;
+
+  enum EventType {
+    UNKNOWN = 0;
+    CHAT_CREATED = 1;
+    NAME_CHANGED = 2;
+    MEMBERS_ADDED = 3;
+    MEMBER_JOINED = 4;
+    MEMBER_REMOVED = 5;
+    ADMINS_ADDED = 6;
+    ADMIN_REMOVED = 7;
+  }
 }
 ```
 
 ### Chat ID
 
-Each membership update MUST be sent with a corresponding `chat-id`. 
-The format of this chat id MUST be a string, [UUID](https://tools.ietf.org/html/rfc4122 ), concatenated with the hex-encoded public key of the creator of the chat. This chat-id MUST be validated by all clients, and MUST be discarded if it does not follow these rules.
+Each membership update MUST be sent with a corresponding `chatId`. 
+The format of this chat id MUST be a string, [UUID](https://tools.ietf.org/html/rfc4122 ), concatenated with the hex-encoded public key of the creator of the chat. This chatId MUST be validated by all clients, and MUST be discarded if it does not follow these rules.
 
 ### Signature
 
-The signature for each event is calculated by creating a JSON array of all the `events` sorted by `clock-value` in ascending order, where each event is transformed in an array of tuples `field-name`, `value`, sorted by `field-name` in ascending alphabetical order. The last element of the array MUST be the `chat-id`. 
-Empty fields MUST be removed.
-
-For example the event:
-
-```
-   {
-     "chat-id": "chat-id",
-       "events":  [
-       {"b": "b-value"
-        "clock-value": 1,
-        "a": "a-value"
-        },
-       {
-         "e": "e-value",
-         "clock-value": 0,
-         "a": "a-value"
-       }
-       ]
-   }
-
-```
-
-Results in the structure:
-
-```
-   [
-     [
-       [
-         ["a" "a-value"],
-         ["clock-value", 0],
-         ["e" "e-value"]
-       ],
-       [
-         ["a", "a-value"],
-         ["b", "b-value"],
-         ["clock-value", 1]
-       ]
-     ],
-     "chat-id"
-   ]
-```
-
-This structure is then stringified collapsing all whitespaces and the `Keccak256` of the string is then signed using its private key by the author and added to the payload.
+The signature for each event is calculated by encoding each `MembershipUpdateEvent` in its protobuf representation
```

Comment:
protobuf is a non-deterministic format - there are many valid serializations for the same data - this will not fly.

Reply:
fwiw,  here's the relevant documentation link: https://developers.google.com/protocol-buffers/docs/encoding#order

Reply:
I don't think that's a problem, I believe we had already a discussion about this at some point.

Basically the fact that protobuf does not guarantees a deterministic format (across versions I believe, same implementation generally is deterministic, but let's assume as well it isn't), does not impact this, for example:

1) I create an event, I encode it to protobuf and sign the bytes generated, send over `signature+bytes`.
2) A client receives it, verify `signature` against `bytes`, decodes the message. 

At no point a client should re-encode the message nor the message should be re-encoded (as well because that's likely to break in protobuf, as unknown fields are simply ignored, so it would break the protocol at the first upgrade), if the message is to be relayed, the original `signature+bytes` should be used, unchanged.

The event itself when it comes to the signature, is treated as just as bytes, the fact that generating those bytes is not deterministic as no impact on the validity of the signature.

If client A (for whatever reason, even if currently there's no use case for this, as re-transmitting uses the raw payload), re-encodes the message twice, and gets two different results, it is also not a problem as the protocol is idempotent so no change will result from duplicates, even if they are not recognized as such.

What do you think?

Reply:
I think this is fishy because it becomes possible to sign byte sequences whose deserialized representation is unexpected - you're signing some random bytes at that point and not a particular message as it logically appears to the receiver - you're proving that you have the private key, but not that it was your intent to sign a particular instruction, which seems to open up the signature for interpretation as well, weakening the value of it.

I agree it's mitigated by the custom (non-protobuf) serialization format in `events` and that the use case is somewhat casual (chat) use case. I anticipate surprises as it starts getting abused.

> unknown fields are simply ignored

not sure what you mean here - it's a feature of protobuf to be able to reencode and alter a message without understanding all the fields - that includes preserving unknown fields (perhhaps not all protobuf implementations do this currently, c++/python did last I checked)

Reply:
What we did in MVDS is order the fields to sign, see `message indentifiers`. https://specs.vac.dev/mvds.html#payloads

Reply:
Not sure I fully understand your point, 
you are not signing some "random bytes", you are signing the encode version of the message you are about to send.
It does imply that you trust the protobuf library to encode the message non maliciously, is that your concern? (that goes to any library that you use to generate the message though, whether you are concatenating the field etc, regardless of the signature scheme)

The method used is the one suggested by the primary author of protobuf https://groups.google.com/d/msg/protobuf/iCA5ALRtDUI/vUiAl6GQqUAJ &  https://capnproto.org/ .

The alternative so far proposed (to sign each field, concatenating etc), are fine with static protocol that don't change often, that's why we used it in the first place in this one, but are more prone to break as newly added field might be ignored by older clients, so won't be used to calculate the signature.

Unkowns fields were ignored in proto 3, seems like they corrected it in proto 3.5 (we still use 3 ourselves):

https://developers.google.com/protocol-buffers/docs/proto3#unknowns

So the method used before (concatenating each field separately), would break when adding a new field (or we have to have a new field with the new signature scheme, maintaining the old one).

What method are you proposing exactly as an alternative?

MVDS strategy for example, given we use protobuf3 , would break if we ever added a new field, and the choice would be, either to exclude it from messageID calculation (which might be fine with messageID maybe, definitely not with signatures), or account that with backward compatible code.

Reply:
Assuming that the reading client parses the data and then displays it to the user as "signed", there could be question about what exactly was signed: was it the parsed message or the byte-stream? I wonder then if the sender can construct a message, sign it and then claim that what they signed was not what the recipient understood it to be (after parsing).

You make a good point that if idempotency is not needed, the serialization matters less - in MVDS the message id is important for the correctness of the sync - if that is not the case here, it's less of a problem than I initially assumed. Thanks.

Reply:
I don't think that would be possible, essentially non-repudiation and integrity still stand, and the result is always unambiguous (we never sign the parsed message but always the byte-stream), even if there's  more than one way to serialize the message (due to the order of fields not being specified/enforced essentially, given `protobuf` is map).

There can be no doubt on the origin and the integrity of the message, whether there might be some doubt on the interpretation it would be a problem in the protocol itself (say for example the two clients have mismatched protobuf specifications).
In general protobuf being not self-describing might lead to such issues, but they would come up regardless of the signature scheme used.

thanks for the feedback

Reply:
This question seems to come up quite frequently, it might be a good idea to add it to a rationale/FAQ section at the end so we don't end up having the same conversation over and over again :)